pjmedia/src/pjmedia-codec/and_aud_mediacodec.cpp

```diff
@@ -0,0 +1,1366 @@
+	info->mode = ((amr_settings_t*)codec_data->codec_setting)->enc_mode;
+	info->start_bit = 0;
+	frames[i].buf = p + 1;
+        if (setting->amr_nb) {
+            frames[i].size = (info->frame_type <= 8)?
+                             pjmedia_codec_amrnb_framelen[info->frame_type] : 0;
+        } else {
+            frames[i].size = (info->frame_type <= 9)?
+                             pjmedia_codec_amrwb_framelen[info->frame_type] : 0;
+        }
+	p += frames[i].size + 1;
+    }
+    /* Pack */
+    *pkt_size = max_pkt_size;
+    status = pjmedia_codec_amr_pack(frames, nframes, setting, pkt, pkt_size);
+
+    return status;
+}
+
+/* Parse AMR payload into frames. */
+static pj_status_t parse_amr(and_media_private_t *codec_data, void *pkt,
+			     pj_size_t pkt_size, const pj_timestamp *ts,
+			     unsigned *frame_cnt, pjmedia_frame frames[])
+{
+    amr_settings_t* s = (amr_settings_t*)codec_data->codec_setting;
+    pjmedia_codec_amr_pack_setting *setting;
+    pj_status_t status;
+    pj_uint8_t cmr;
+
+    setting = &s->dec_setting;
+    status = pjmedia_codec_amr_parse(pkt, pkt_size, ts, setting, frames, 
+				     frame_cnt, &cmr);
+    if (status != PJ_SUCCESS)
+	return status;
+
+    /* Check Change Mode Request. */
+    if (((setting->amr_nb && cmr <= 7) || (!setting->amr_nb && cmr <= 8)) &&
+	s->enc_mode != cmr)
+    {
+	s->enc_mode = cmr;
+    }
+    return PJ_SUCCESS;
+}
+
+static void predecode_amr(and_media_private_t *codec_data,
+			  const pjmedia_frame *input,
+			  pjmedia_frame *out)
+{
+    pjmedia_codec_amr_bit_info *info;
+    pj_uint8_t *bitstream = (pj_uint8_t *)out->buf;
+    pjmedia_codec_amr_pack_setting *setting;
+
+    out->buf = &bitstream[1];
+    setting = &((amr_settings_t*)codec_data->codec_setting)->dec_setting;
+    pjmedia_codec_amr_predecode(input, setting, out);
+    info = (pjmedia_codec_amr_bit_info*)&out->bit_info;
+    bitstream[0] = (info->frame_type << 3) | (info->good_quality << 2);
+    out->buf = &bitstream[0];
+    ++out->size;
+}
+
+#endif /* PJMEDIA_HAS_AND_MEDIA_AMRNB || PJMEDIA_HAS_AND_MEDIA_AMRWB */
+
+static pj_status_t configure_codec(and_media_private_t *and_media_data,
+				   pj_bool_t is_encoder)
+{
+    media_status_t am_status;
+    AMediaFormat *aud_fmt;
+    int idx = and_media_data->codec_idx;
+    AMediaCodec *codec = (is_encoder?and_media_data->enc:and_media_data->dec);
+
+    aud_fmt = AMediaFormat_new();
+    if (!aud_fmt) {
+        return PJ_ENOMEM;
+    }
+    AMediaFormat_setString(aud_fmt, AND_MEDIA_KEY_MIME,
+                           and_media_codec[idx].mime_type);
+    AMediaFormat_setInt32(aud_fmt, AND_MEDIA_KEY_PCM_ENCODING, 2);
+    AMediaFormat_setInt32(aud_fmt, AND_MEDIA_KEY_CHANNEL_COUNT,
+                          and_media_codec[idx].channel_count);
+    AMediaFormat_setInt32(aud_fmt, AND_MEDIA_KEY_SAMPLE_RATE,
+			  and_media_codec[idx].clock_rate);
+    AMediaFormat_setInt32(aud_fmt, AND_MEDIA_KEY_BITRATE,
+			  and_media_codec[idx].def_bitrate);
```

Comment:
If there are any settings for VAD, DTX, PLC, etc, better configure them here too.

Btw, can we enable/disable VAD/PLC for AMR/AMR-WB? If not, the docs better mention it.

Reply:
There's no VAD, DTX, PLC feature from MediaCodec. Add documentation on ```and_aud_mediacodec.h```